metrics/engine/engine.go

```diff
@@ -0,0 +1,174 @@
+	// TODO: replace with strings.Cut after Go 1.18
+	nameParts := strings.SplitN(name, "{", 2)
+
+	metric := me.registry.Get(nameParts[0])
+	if metric == nil {
+		return nil, fmt.Errorf("metric '%s' does not exist in the script", nameParts[0])
+	}
+	if len(nameParts) == 1 { // no sub-metric
+		return metric, nil
+	}
+
+	if nameParts[1][len(nameParts[1])-1] != '}' {
+		return nil, fmt.Errorf("missing ending bracket, sub-metric format needs to be 'metric{key:value}'")
+	}
+	sm, err := metric.AddSubmetric(nameParts[1][:len(nameParts[1])-1])
+	if err != nil {
+		return nil, err
+	}
+	return sm.Metric, nil
+}
+
+func (me *MetricsEngine) markObserved(metric *stats.Metric) {
+	if !metric.Observed {
+		metric.Observed = true
+		me.ObservedMetrics[metric.Name] = metric
+	}
+}
+
+func (me *MetricsEngine) initSubMetricsAndThresholds() error {
+	for metricName, thresholds := range me.options.Thresholds {
+		metric, err := me.getOrInitPotentialSubmetric(metricName)
+
+		if me.runtimeOptions.NoThresholds.Bool {
+			if err != nil {
+				me.logger.WithError(err).Warnf("Invalid metric '%s' in threshold definitions", metricName)
+			}
+			continue
+		}
+
+		if err != nil {
+			return fmt.Errorf("invalid metric '%s' in threshold definitions: %w", metricName, err)
+		}
+
+		metric.Thresholds = thresholds
+		me.metricsWithThresholds = append(me.metricsWithThresholds, metric)
+
+		// Mark the metric (and the parent metric, if we're dealing with a
+		// submetric) as observed, so they are shown in the end-of-test summary,
+		// even if they don't have any metric samples during the test run
+		me.markObserved(metric)
+		if metric.Sub != nil {
+			me.markObserved(metric.Sub.Metric)
+		}
+	}
+
+	// TODO: refactor out of here when https://github.com/grafana/k6/issues/1321
+	// lands and there is a better way to enable a metric with tag
+	if me.options.SystemTags.Has(stats.TagExpectedResponse) {
+		_, err := me.getOrInitPotentialSubmetric("http_req_duration{expected_response:true}")
+		if err != nil {
+			return err // shouldn't happen, but ¯\_(ツ)_/¯
+		}
+	}
+
+	return nil
+}
+
+// ProcessThresholds processes all of the thresholds.
+//
+// TODO: refactor, make private, optimize
+func (me *MetricsEngine) ProcessThresholds() (thresholdsTainted, shouldAbort bool) {
+	me.MetricsLock.Lock()
+	defer me.MetricsLock.Unlock()
+
+	t := me.executionState.GetCurrentTestRunDuration()
+
+	for _, m := range me.metricsWithThresholds {
+		if len(m.Thresholds.Thresholds) == 0 {
+			continue
+		}
+		m.Tainted = null.BoolFrom(false)
+
+		me.logger.WithField("m", m.Name).Debug("running thresholds")
+		succ, err := m.Thresholds.Run(m.Sink, t)
+		if err != nil {
+			me.logger.WithField("m", m.Name).WithError(err).Error("Threshold error")
+			continue
+		}
+		if !succ {
+			me.logger.WithField("m", m.Name).Debug("Thresholds failed")
+			m.Tainted = null.BoolFrom(true)
+			thresholdsTainted = true
+			if m.Thresholds.Abort {
+				shouldAbort = true
+			}
+		}
```

Comment:
```suggestion
		if succ {
		   continue
		}
		
		me.logger.WithField("m", m.Name).Debug("Thresholds failed")
		m.Tainted = null.BoolFrom(true)
		thresholdsTainted = true
		if m.Thresholds.Abort {
			shouldAbort = true
		}		
```
:see_no_evil: sorry, I really don't like nested `if`s

Reply:
resolved by https://github.com/grafana/k6/pull/2426/commits/d7bc04b33223b6c00f944c6d57b82bdba4bef52f